#  ***Proyecto* *de* *Métodos* *Cuantitativo***
Pronóstico del clima en Kansas
# Estudiantes

Chaves Mora, Aldahir - B92175     
Rosales Mora, Esteban - B96967                 
Sandi Calderon, María Fernanda - B77168       
Vasquez, Erick- B98334        
Zúñiga Díaz, Alan Fabián - B78597



In [3]:
import numpy as np
import pandas as pd
import matplotlib as npl
import scipy as sp
import seaborn as sns
import statsmodels.formula.api as smf

In [4]:
dataKansas = pd.read_csv("KANSAS_CITY_INTERNATIONAL _AIRPORT.csv")

Vamos a visualizar los datos de la ciudad Kansas 

In [5]:
dataKansas

STATION                                      NAME        DATE  \
0     USW00003947  KANSAS CITY INTERNATIONAL AIRPORT, MO US  2009-01-02   
1     USW00003947  KANSAS CITY INTERNATIONAL AIRPORT, MO US  2009-01-03   
2     USW00003947  KANSAS CITY INTERNATIONAL AIRPORT, MO US  2009-01-04   
3     USW00003947  KANSAS CITY INTERNATIONAL AIRPORT, MO US  2009-01-05   
4     USW00003947  KANSAS CITY INTERNATIONAL AIRPORT, MO US  2009-01-06   
...           ...                                       ...         ...   
4907  USW00003947  KANSAS CITY INTERNATIONAL AIRPORT, MO US  2022-06-10   
4908  USW00003947  KANSAS CITY INTERNATIONAL AIRPORT, MO US  2022-06-11   
4909  USW00003947  KANSAS CITY INTERNATIONAL AIRPORT, MO US  2022-06-12   
4910  USW00003947  KANSAS CITY INTERNATIONAL AIRPORT, MO US  2022-06-13   
4911  USW00003947  KANSAS CITY INTERNATIONAL AIRPORT, MO US  2022-06-14   

       AWND    FMTM    PGTM  PRCP  SNOW  SNWD  TAVG  TMAX  TMIN   WDF2   WDF5  \
0      8.28  1047.0  1047.0  0.00   0.0   0.0   NaN    50    23  150.0  150.0   
1     12.53  2020.0  1732.0  0.00   0.0   0.0   NaN    63    22  330.0  320.0   
2     10.51   403.0   402.0  0.00   0.0   0.0   NaN    25    14  340.0  340.0   
3      6.26  1907.0  1552.0  0.00   0.0   0.0   NaN    35    16  140.0  210.0   
4      5.14    58.0    51.0  0.00   0.0   0.0   NaN    39    23  130.0  130.0   
...     ...     ...     ...   ...   ...   ...   ...   ...   ...    ...    ...   
4907    NaN     NaN     NaN  0.11   0.0   0.0  73.0    85    64    NaN    NaN   
4908    NaN     NaN     NaN  0.00   0.0   0.0  77.0    89    65    NaN    NaN   
4909    NaN     NaN     NaN  0.00   0.0   0.0  82.0    91    74    NaN    NaN   
4910    NaN     NaN     NaN  0.00   0.0   0.0  86.0    97    81    NaN    NaN   
4911    NaN     NaN     NaN  0.00   0.0   0.0   NaN    93    81    NaN    NaN   

      WSF2  WSF5  
0     19.9  23.0  
1     25.9  34.9  
2     23.9  30.0  
3     13.0  23.9  
4     10.1  14.1  
...    ...   ...  
4907   NaN   NaN  
4908   NaN   NaN  
4909   NaN   NaN  
4910   NaN   NaN  
4911   NaN   NaN  

[4912 rows x 16 columns]

Dimencion de los datos de Kansas                      
En donde tiene 4912 filas x 16 columnas 

In [6]:
dataKansas.shape

(4912, 16)

Es importante conocer los datos ausentes del modelo que se esta entrenado, en donde como podemos ver la mayoria de las variables no tiene datos ausentes.

In [7]:
dataKansas.isna().sum().sort_values()

STATION       0
NAME          0
DATE          0
PRCP          0
SNOW          0
TMAX          0
TMIN          0
SNWD          2
AWND          6
WDF2          6
WSF2          6
WDF5         25
WSF5         25
TAVG       1551
FMTM       3879
PGTM       3913
dtype: int64

Ahora bien, para realizar un entrenamiento de datos es necesario quitar los datos faltantes, de la BD que estamos entrenando.                                 
Por ello vamos aponer 0 en donde hay datos nulos.

In [8]:
dataKansas = dataKansas.fillna(0)
print("Tabla con datos eliminados")
print(dataKansas)

Tabla con datos eliminados
          STATION                                      NAME        DATE  \
0     USW00003947  KANSAS CITY INTERNATIONAL AIRPORT, MO US  2009-01-02   
1     USW00003947  KANSAS CITY INTERNATIONAL AIRPORT, MO US  2009-01-03   
2     USW00003947  KANSAS CITY INTERNATIONAL AIRPORT, MO US  2009-01-04   
3     USW00003947  KANSAS CITY INTERNATIONAL AIRPORT, MO US  2009-01-05   
4     USW00003947  KANSAS CITY INTERNATIONAL AIRPORT, MO US  2009-01-06   
...           ...                                       ...         ...   
4907  USW00003947  KANSAS CITY INTERNATIONAL AIRPORT, MO US  2022-06-10   
4908  USW00003947  KANSAS CITY INTERNATIONAL AIRPORT, MO US  2022-06-11   
4909  USW00003947  KANSAS CITY INTERNATIONAL AIRPORT, MO US  2022-06-12   
4910  USW00003947  KANSAS CITY INTERNATIONAL AIRPORT, MO US  2022-06-13   
4911  USW00003947  KANSAS CITY INTERNATIONAL AIRPORT, MO US  2022-06-14   

       AWND    FMTM    PGTM  PRCP  SNOW  SNWD  TAVG  TMAX  TMIN   WDF2  

Luego para conocer el analisis descriptivo de los datosde nuestra BD

In [9]:
dataKansas.describe()


AWND         FMTM         PGTM         PRCP         SNOW  \
count  4912.000000  4912.000000  4912.000000  4912.000000  4912.000000   
mean     10.030719   336.844870   272.836726     0.112028     0.053888   
std       3.773768   972.893327   604.363002     0.352062     0.440781   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       7.380000     0.000000     0.000000     0.000000     0.000000   
50%       9.400000     0.000000     0.000000     0.000000     0.000000   
75%      12.135000     0.000000     0.000000     0.020000     0.000000   
max      26.840000  9999.000000  2359.000000     4.480000     9.200000   

              SNWD         TAVG         TMAX         TMIN         WDF2  \
count  4912.000000  4912.000000  4912.000000  4912.000000  4912.000000   
mean      0.192529    37.955619    65.314129    44.982899   192.516287   
std       0.948576    30.336202    20.441295    19.527354   101.571944   
min       0.000000    -5.000000    -1.000000   -13.000000     0.000000   
25%       0.000000     0.000000    50.000000    30.000000   120.000000   
50%       0.000000    42.000000    68.000000    46.000000   190.000000   
75%       0.000000    67.000000    83.000000    63.000000   280.000000   
max      10.000000    89.000000   107.000000    81.000000   360.000000   

              WDF5         WSF2         WSF5  
count  4912.000000  4912.000000  4912.000000  
mean    195.171010    20.063864    26.308571  
std     102.162981     6.566619     9.013347  
min       0.000000     0.000000     0.000000  
25%     120.000000    15.000000    19.900000  
50%     190.000000    19.900000    25.100000  
75%     290.000000    23.900000    32.000000  
max     360.000000    53.000000    67.100000

Entrenar el modelo

In [10]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression 
from sklearn.linear_model import Ridge

In [11]:
X_train, X_test, y_train, y_test = train_test_split( dataKansas.drop(['AWND', 
                                          'PRCP', 'SNWD'], axis = 'columns'),
                                        dataKansas['TMAX'],
                                        train_size   = 0.8,
                                        random_state = 1234,
                                        shuffle      = True
                                    )

In [12]:
print("Partición de entrenamento")
print(" ")
print(y_train.describe())

print(" ")

print("Partición de test")
print("")
print(y_test.describe())


Partición de entrenamento
 
count    3929.000000
mean       65.152202
std        20.616831
min        -1.000000
25%        50.000000
50%        68.000000
75%        82.000000
max       107.000000
Name: TMAX, dtype: float64
 
Partición de test

count    983.000000
mean      65.961343
std       19.721031
min        5.000000
25%       52.000000
50%       69.000000
75%       83.000000
max      101.000000
Name: TMAX, dtype: float64


In [13]:

from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.compose import make_column_selector

numeric_cols = X_train.select_dtypes(include=['float64', 'int']).columns.to_list()
cat_cols = X_train.select_dtypes(include=['object', 'category']).columns.to_list()

preprocessor = ColumnTransformer(
                   [('scale', StandardScaler(), numeric_cols),
                    ('onehot', OneHotEncoder(handle_unknown='ignore'), cat_cols)],
                remainder='passthrough')

In [14]:
X_train_prep = preprocessor.fit_transform(X_train)
X_test_prep  = preprocessor.transform(X_test)

In [15]:
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.compose import make_column_selector

numeric_cols = X_train.select_dtypes(include=['float64', 'int']).columns.to_list()
cat_cols = X_train.select_dtypes(include=['object', 'category']).columns.to_list()


numeric_transformer = Pipeline(
                        steps=[
                            ('imputer', SimpleImputer(strategy='median')),
                            ('scaler', StandardScaler())
                        ]
                      )


categorical_transformer = Pipeline(
                            steps=[
                                ('imputer', SimpleImputer(strategy='most_frequent')),
                                ('onehot', OneHotEncoder(handle_unknown='ignore'))
                            ]
                          )

preprocessor = ColumnTransformer(
                    transformers=[
                        ('numeric', numeric_transformer, numeric_cols),
                        ('cat', categorical_transformer, cat_cols)
                    ],
                    remainder='passthrough'
                )

In [16]:
X_train_prep = preprocessor.fit_transform(X_train)
X_test_prep  = preprocessor.transform(X_test)

In [17]:
numeric_cols = X_train.select_dtypes(include=['float64', 'int']).columns.to_list()
cat_cols = X_train.select_dtypes(include=['object', 'category']).columns.to_list()

numeric_transformer = Pipeline(
                        steps=[('scaler', StandardScaler())]
                      )

categorical_transformer = Pipeline(
                            steps=[('onehot', OneHotEncoder(handle_unknown='ignore'))]
                          )

preprocessor = ColumnTransformer(
                    transformers=[
                        ('numeric', numeric_transformer, numeric_cols),
                        ('cat', categorical_transformer, cat_cols)
                    ],
                    remainder='passthrough'
                )

pipe = Pipeline([('preprocessing', preprocessor),
                 ('modelo', Ridge())])


_ = pipe.fit(X=X_train, y=y_train)

In [18]:

from sklearn.model_selection import cross_val_score

cv_scores = cross_val_score(
                estimator = pipe,
                X         = X_train,
                y         = y_train,
                scoring   = 'neg_root_mean_squared_error',
                cv        = 5
             )

print(f"Métricas validación cruzada: {cv_scores}")
print(f"Média métricas de validación cruzada: {cv_scores.mean()}")

Métricas validación cruzada: [-0.03647135 -0.03918891 -0.03843654 -0.03932329 -0.04210796]
Média métricas de validación cruzada: -0.03910561049014308


In [19]:
from sklearn.model_selection import RepeatedKFold
from sklearn.model_selection import cross_validate

cv = RepeatedKFold(n_splits=3, n_repeats=5, random_state=123)
cv_scores = cross_validate(
                estimator = pipe,
                X         = X_train,
                y         = y_train,
                scoring   = ('r2', 'neg_root_mean_squared_error'),
                cv        = cv,
                return_train_score = True
            )


cv_scores = pd.DataFrame(cv_scores)
cv_scores

fit_time  score_time   test_r2  train_r2  \
0   0.031418    0.015062  0.999995  0.999999   
1   0.028331    0.013872  0.999995  0.999999   
2   0.028091    0.013510  0.999995  0.999999   
3   0.026524    0.014215  0.999995  0.999999   
4   0.029651    0.013699  0.999995  0.999999   
5   0.035561    0.013733  0.999995  0.999999   
6   0.037102    0.013798  0.999995  0.999999   
7   0.027404    0.014048  0.999995  0.999999   
8   0.026640    0.013854  0.999994  0.999998   
9   0.026891    0.014050  0.999995  0.999999   
10  0.028014    0.014138  0.999995  0.999999   
11  0.026510    0.013535  0.999996  0.999999   
12  0.026529    0.013825  0.999995  0.999999   
13  0.028988    0.015494  0.999995  0.999999   
14  0.027086    0.013651  0.999995  0.999999   

    test_neg_root_mean_squared_error  train_neg_root_mean_squared_error  
0                          -0.046086                          -0.023715  
1                          -0.044446                          -0.022014  
2                          -0.043935                          -0.022244  
3                          -0.045210                          -0.024282  
4                          -0.045368                          -0.024580  
5                          -0.044604                          -0.022261  
6                          -0.044049                          -0.022152  
7                          -0.044828                          -0.023123  
8                          -0.047746                          -0.027357  
9                          -0.045688                          -0.022644  
10                         -0.045047                          -0.023856  
11                         -0.042896                          -0.021511  
12                         -0.045729                          -0.024132  
13                         -0.044816                          -0.023791  
14                         -0.044633                          -0.022194

In [20]:
predicciones = pipe.predict(X_test)

In [21]:
df_predicciones = pd.DataFrame({'TMAX' : y_test, 'prediccion' : predicciones})
df_predicciones.head()

TMAX  prediccion
3186    88   87.987088
3696    17   17.078246
1303    95   94.992499
3141    83   82.982493
4452    50   50.062587

In [22]:
from sklearn.metrics import mean_squared_error

rmse = mean_squared_error(
        y_true = y_test,
        y_pred = predicciones,
        squared = False
       )
rmse

0.03408528656996363